distributed3/pool.py

```diff
@@ -0,0 +1,475 @@
+
+    def map(self, func, seq, **kwargs):
+        return IOLoop.current().run_sync(
+                lambda: self._map(func, seq, **kwargs))
+
+    def sync_center(self):
+        """ Get who_has and has_what dictionaries from a center
+
+        In particular this tells us what workers we have at our disposal
+        """
+        return IOLoop.current().run_sync(self._sync_center)
+
+    @gen.coroutine
+    def _close_connections(self):
+        """ Close active connections """
+        for stream in self._open_streams:
+            if not stream._closed:
+                r = rpc(stream)
+                result = yield r.close(close=True)
+        yield self.center.close(close=True)
+        self.center.close_streams()
+
+    def close_connections(self):
+        return IOLoop.current().run_sync(self._close_connections)
+
+    def close(self):
+        """ Close the thread that manages our event loop """
+        self.close_connections()
+
+    @gen.coroutine
+    def _apply_async(self, func, args=(), kwargs={}, key=None):
+        if not isinstance(args, (tuple, list)):
+            raise TypeError('args must be a tuple as in:\n'
+                    '  pool.apply_async(func, args=(x,))')
+        [args2, kwargs2], needed = unpack_remotedata([args, kwargs])
+
+        ip, port = choose_worker(needed, self.who_has, self.has_what,
+                                 self.available_cores)
+
+        if key is None:
+            key = str(uuid.uuid1())
+
+        pc = PendingComputation(key, func, args2, kwargs2, needed,
+                                self.center.ip, self.center.port)
+        yield pc._start(ip, port, self.who_has, self.has_what,
+                        self.available_cores)
+        raise Return(pc)
+
+    def apply_async(self, func, args=(), kwargs={}, key=None):
+        """ Execute a function in a remote worker
+
+        If an arg or a kwarg is a ``RemoteData`` object then that data will be
+        communicated as necessary among the ``Worker`` peers.
+        """
+        return IOLoop.current().run_sync(
+                lambda: self._apply_async(func, args, kwargs, key))
+
+    def apply(self, func, args=(), kwargs={}, key=None):
+        return self.apply_async(func, args, kwargs, key).get()
+
+    @gen.coroutine
+    def _scatter(self, data, key=None):
+        yield self._sync_center()
+        result = yield scatter_to_workers(self.center.ip,
+                self.center.port, self.ncores, data, key=key)
+        raise Return(result)
+
+    def scatter(self, data, key=None):
+        return IOLoop.current().run_sync(
+                lambda: self._scatter(data, key))
+
+    @gen.coroutine
+    def _gather(self, data):
+        result = yield gather_from_center((self.center.ip, self.center.port),
+                                          data)
+        raise Return(result)
+
+    def gather(self, data):
+        return IOLoop.current().run_sync(lambda: self._gather(data))
+
+
+class PendingComputation(object):
```

Comment:
When I do things like this, I like to match existing APIs, e.g. subclass concurrent.Future or match multiprocessing.pool.AsyncResult. Future probably makes the most sense here - that way people familiar with tornado coroutines are already familiar with how to talk to your async APIs.